### PR TITLE
Reduce Metadata index deletes and writes

### PR DESCRIPTION
**Pull request by dwedul-figure**
## Description

Reduce the number of deletes and sets done for indexes by comparing the old and new versions of Metadata, and only doing what's needed.

closes: #543

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Review by iramiller — DISMISSED** [review]
Reviewing the amount of effort we have in building query indexes... as well as the amount of merkle tree pressure that results from these values makes me realize that the split off of queries from state into separate databases is extremely important for us.  This update should go a long way in mitigating the kvstore write pressure of minor scope changes.

**Review by iramiller — APPROVED** [review]

File: x/metadata/keeper/scope.go [hunk 1]
```diff
@@ -127,66 +129,109 @@ func (k Keeper) RemoveScope(ctx sdk.Context, id types.MetadataAddress) {
...
 	defer types.GetIncObjFunc(types.TLType_Scope, types.TLAction_Deleted)
 }
 
-// clearScopeIndex delete any index records for this scope
-func (k Keeper) clearScopeIndex(ctx sdk.Context, scope types.Scope) {
-	store := ctx.KVStore(k.storeKey)
+// scopeIndexValues is a struct containing the values used to index a scope.
+type scopeIndexValues struct {
+	ScopeID         types.MetadataAddress
+	Addresses       []string
```
Comment (arnabmitra) [inline x/metadata/keeper/scope.go:141]: curious why not typesafety for addresses and valueowner?

**Review by arnabmitra — APPROVED** [review]
lgtm

File: x/metadata/keeper/scope.go (hunk 1 above)
Comment (dwedul-figure) [reply x/metadata/keeper/scope.go:141]: I just used the same types that they are in the scopes. Same with the other similar `*IndexValues` structs.

File: x/metadata/keeper/scope.go [hunk 2]
```diff
@@ -127,66 +129,109 @@ func (k Keeper) RemoveScope(ctx sdk.Context, id types.MetadataAddress) {
...
+
+// indexScope updates the index entries for a scope.
+//
+// When adding a new scope:  indexScope(ctx, scope, nil)
+//
+// When deleting a scope:  indexScope(ctx, nil, scope)
+//
+// When updating a scope:  indexScope(ctx, newScope, oldScope)
+//
+// If both newScope and oldScope are not nil, it is assumed that they have the same ScopeId.
```
[the remark this answers is not in the document]
Comment (dwedul-figure) [reply x/metadata/keeper/scope.go:216]: Nope. This part of the comment is pointing out an assumption that was made for cases when both an `newScope` and `oldScope` are provided, e.g. "When updating a scope."

**Pull request merged**